### PR TITLE
Reader sidebar: toggle visibility of lists and tags

### PR DESCRIPTION
**Pull request by bluefuton**
Rather than showing all of the user's lists and tags in the sidebar, this PR experiments with hiding them behind a toggle.

Default state:

screen shot 2015-11-25 at 11 31 17

Expanded:

screen shot 2015-11-25 at 11 31 27

cc @shaunandrews

### CODE REVIEW

**Comment by bluefuton** [comment]
As a starting point, I think I'd create a common component called something like `SidebarFoldableSection` (perhaps based on `FoldableCard`) which handles the open/close state and can be used for both tags and lists.

**Comment by bluefuton** [comment]
I don't think we'll be able to dynamically update the count on lists until we've got a ReaderListStore in place (part of https://github.com/Automattic/wp-calypso/milestones/Reader:%20List%20Management).

**Comment by bluefuton** [comment]
Hey @shaunandrews - in e617ccb I added a new `SidebarMenu` component that manages its own open/closed state, just like `FoldableCard` does. 

That means we don't need to keep track of the open state in the sidebar itself, which should clean things up a bit. We can also use the same component to render the lists and tags menus, which should reduce duplication of logic.

The new component doesn't include the 'add new list/tag' input yet, so that'd be the next step :)

(Sidenote: I had to rebase the branch and force push to get the dependencies working, so it's worth doing a `git fetch && git reset --hard origin/add/sidebar-section-toggle` to grab my changes.)

**Comment by shaunandrews** [comment]
Awesome, @bluefuton. I've brought over the add button and input and things seem to mostly work for Lists, but adding a Tag is broken. Removing a tag is broken — not sure how to fix that. We also need to find a way to remember the state of the toggles for users.

**Comment by bluefuton** [comment]
Ah - I think the tags were failing because of 

`React.findDOMNode( this.refs.menuAddInput ).value = '';`

That input is now in the MenuSidebar component, so it's not in refs any more.

This wasn't tripping up lists because we were doing a redirect to Atlas before the line above got executed :)

I've added it to the SidebarMenu component now and it seems to be working well.

**Comment by bluefuton** [comment]
> We also need to find a way to remember the state of the toggles for users.

Yep. If we want to remember it, we should store it in the state of the parent component, `ReaderSidebar`.

**Comment by shaunandrews** [comment]
@bluefuton I rebased this and somewhere along the way I broke the add stuff again. Mind taking another look?

**Comment by bluefuton** [comment]
@shaunandrews Absolutely, I'll take a look. There have been a few sidebar changes happening as part of List Management, so one of those is probably the culprit...

**Comment by ralder** [comment]
There are few issues what i found inside 'menu.jsx':
- in function [handleAddKeyDown](https://github.com/Automattic/wp-calypso/pull/724/files#diff-7d53f989287e53870743ebe25af67ecbR76) missed input parameter `event` 
- also in this function not handled case if value of text input is empty and i press enter
- since React 0.14 not need using findDOMNode 
- also will be good if when i press Add button text input get focus
- for placeholder and button title need use `translate`

**Comment by bluefuton** [comment]
Thanks for the feedback @ralder!

I've addressed your points in 19629f7. And as a result the add input is now working again, @shaunandrews :)

**Comment by bluefuton** [comment]
Todo: 
- [ ] persist open and closed state between page views

**Comment by bluefuton** [comment]
Styling of the lists and tags menu is broken after a non-very-fun rebase :expressionless: The toggle is still functional though.

**Comment by bluefuton** [comment]
Well, this turned into quite the :japanese_ogre: 

...but includes a long overdue refactoring of the sidebar, and the first sprinkling of Redux in the Reader.

**Comment by bluefuton** [comment]
cc @blowery @shaunandrews

File: client/layout/sidebar/menu.jsx [hunk 1]
```diff
@@ -2,9 +2,10 @@
  * External dependencies
  */
 import React from 'react';
+import classNames from 'classnames';
 
-const SidebarMenu = ( { children } ) => (
-	<li className="sidebar__menu">{ children }</li>
+const SidebarMenu = ( { children, className } ) => (
```
Comment (blowery) [inline client/layout/sidebar/menu.jsx]: hmmmm... do we _really_ need this?

File: client/reader/controller.js [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
  * External Dependencies
  */
-var ReactDom = require( 'react-dom' ),
+const ReactDom = require( 'react-dom' ),
 	React = require( 'react' ),
 	page = require( 'page' ),
 	debug = require( 'debug' )( 'calypso:reader:controller' ),
-	trim = require( 'lodash/string/trim' );
+	trim = require( 'lodash/string/trim' ),
+	ReduxProvider = require( 'react-redux' ).Provider;
```
Comment (blowery) [inline client/reader/controller.js]: I think you can just call this Provider. That's how it usually appears.

File: client/reader/sidebar/_style.scss [hunk 3]
```diff
@@ -31,55 +34,198 @@
 		}
 	}
 
-	.sidebar-streams {
-	}
+	.sidebar-streams {}
```
Comment (blowery) [inline client/reader/sidebar/_style.scss]: can just remove this

File: client/reader/sidebar/_style.scss [hunk 4]
```diff
@@ -31,55 +34,198 @@
...
+			padding: 10px 16px 10px 55px;
+			cursor: pointer;
+			position: relative;
+			transition: background-color 0.15s ease-in-out,
+				box-shadow 0.15s ease-in-out;
+
+			.gridicon {
+				top: 10px;
+				left: 20px;
+				height: 16px; // I know, I know...
```
Comment (blowery) [inline client/reader/sidebar/_style.scss]: ?

Comment (bluefuton) [inline client/reader/sidebar/_style.scss]: Ask @shaunandrews :)

File: client/reader/sidebar/expandable.jsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	renderContent() {
+		return (
+			<ul className="sidebar__menu-list">
+				{ this.props.children }
+			</ul>
+		);
+	},
+
+	renderHeader() {
+		const headerClasses = classNames( 'sidebar__menu-header' );
```
Comment (blowery) [inline client/reader/sidebar/expandable.jsx]: don't really need classNames here...

File: client/reader/sidebar/expandable.jsx [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		const inputValue = this.refs.menuAddInput.value;
+		if ( event.keyCode === 13 && inputValue.length > 0 ) {
+			event.preventDefault();
+			this.props.onAddSubmit( inputValue );
+			this.refs.menuAddInput.value = '';
+			this.toggleAdd();
+		}
+	},
+
+	renderAdd() {
```
Comment (blowery) [inline client/reader/sidebar/expandable.jsx]: maybe make this it's own component?

File: client/layout/sidebar/menu.jsx (hunk 1 above)
Comment (bluefuton) [inline client/layout/sidebar/menu.jsx]: The component in general?

File: client/reader/sidebar/expandable.jsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		if ( this.props.disabled ) {
+			return;
+		}
+
+		if ( this.props.onClick ) {
+			this.props.onClick();
+		}
+	},
+
+	renderContent() {
```
Comment (blowery) [inline client/reader/sidebar/expandable.jsx]: i don't think this needs it's own function. inline is fine.

File: client/reader/sidebar/expandable.jsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		return (
+			<ul className="sidebar__menu-list">
+				{ this.props.children }
+			</ul>
+		);
+	},
+
+	renderHeader() {
```
Comment (blowery) [inline client/reader/sidebar/expandable.jsx]: maybe this should be it's own component?

File: client/reader/sidebar/expandable.jsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+		const headerClasses = classNames( 'sidebar__menu-header' );
+		return (
+			<div className={ headerClasses } onClick={ this.onClick }>
+				<SidebarHeading>
+					<Gridicon icon="chevron-down" />
+					<span>{ this.props.title }</span>
+					<Count count={ this.props.count } />
+				</SidebarHeading>
+
+				<div></div>
```
Comment (blowery) [inline client/reader/sidebar/expandable.jsx]: what's this for?

File: client/reader/sidebar/expandable.jsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+			<ul className="sidebar__menu-list">
+				{ this.props.children }
+			</ul>
+		);
+	},
+
+	renderHeader() {
+		const headerClasses = classNames( 'sidebar__menu-header' );
+		return (
+			<div className={ headerClasses } onClick={ this.onClick }>
```
Comment (blowery) [inline client/reader/sidebar/expandable.jsx]: Can we avoid the wrapper div and pass classNames and an onClick to SidebarHeading?

File: client/reader/sidebar/package.json [hunk 11]
```diff
@@ -1,6 +0,0 @@
-{
-  "name": "reader sidebar",
-  "version": "0.0.0",
-  "private": true,
-  "main": "sidebar.jsx"
```
Comment (blowery) [inline client/reader/sidebar/package.json]: yassssssssssssss

File: client/reader/sidebar/_style.scss (hunk 4 above)
Comment (shaunandrews) [inline client/reader/sidebar/_style.scss]: Well, 16px may be a little smaller than what's normally recommended for a gridicon. "I know, I know..."

File: client/reader/controller.js (hunk 2 above)
Comment (bluefuton) [inline client/reader/controller.js]: I noticed it's called `ReduxProvider` in our other controllers, so I stuck with that:

https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=reduxprovider

File: client/reader/sidebar/_style.scss (hunk 3 above)
Comment (bluefuton) [inline client/reader/sidebar/_style.scss]: :+1:

File: client/reader/sidebar/_style.scss (hunk 4 above)
Comment (bluefuton) [inline client/reader/sidebar/_style.scss]: I've changed to "Smaller than recommended for gridicon, but works here" :)

File: client/reader/sidebar/expandable.jsx (hunk 5 above)
Comment (bluefuton) [inline client/reader/sidebar/expandable.jsx]: +1

File: client/reader/sidebar/expandable.jsx (hunk 9 above)
Comment (bluefuton) [inline client/reader/sidebar/expandable.jsx]: :+1: s'gone.

File: client/reader/sidebar/package.json (hunk 11 above)
Comment (bluefuton) [inline client/reader/sidebar/package.json]: :tada:

File: client/reader/sidebar/expandable.jsx (hunk 7 above)
Comment (bluefuton) [inline client/reader/sidebar/expandable.jsx]: :+1:

File: client/reader/sidebar/expandable.jsx (hunk 10 above)
Comment (bluefuton) [inline client/reader/sidebar/expandable.jsx]: :+1:

File: client/reader/sidebar/expandable.jsx (hunk 8 above)
Comment (bluefuton) [inline client/reader/sidebar/expandable.jsx]: :+1: Broken out to `ExpandableSidebarHeading`.

File: client/reader/sidebar/expandable.jsx (hunk 6 above)
Comment (bluefuton) [inline client/reader/sidebar/expandable.jsx]: :+1: Broken out to `ExpandableSidebarAddForm`.

File: client/reader/controller.js (hunk 2 above)
Comment (blowery) [inline client/reader/controller.js]: Oh, interesting. That works.

File: client/layout/sidebar/menu.jsx (hunk 1 above)
Comment (blowery) [inline client/layout/sidebar/menu.jsx]: The class names on the component. What are we using it for? Do we need the class name?

**Comment by blowery** [comment]
Reading a tag or a list, then reloading the page, the selected list or tag is not expanded and selected. We'll need to expand a sidebar with a picked item and make sure the item has scrolled into view.

**Comment by blowery** [comment]
Expansion state isn't saved across page loads, and it probably should be. Maybe just stash it in local storage.

File: client/layout/sidebar/menu.jsx (hunk 1 above)
Comment (bluefuton) [inline client/layout/sidebar/menu.jsx]: It's the parent element inside `ExpandableSidebarMenu`, which passes in classes based on the toggle state:

```		const classes = classNames(
			this.props.className,
			{
				'is-toggle-open': !! this.props.expanded,
				'is-togglable': true
			}
		);```

**Comment by bluefuton** [comment]
> Expansion state isn't saved across page loads, and it probably should be. Maybe just stash it in local storage.

It should currently be preserved while moving around in Calypso, but not after a refresh. Maybe one for another PR?

We should do with a general approach for popping parts of the Redux state into local storage, perhaps https://github.com/elgerlambert/redux-localstorage?

**Comment by bluefuton** [comment]
> Reading a tag or a list, then reloading the page, the selected list or tag is not expanded and selected. We'll need to expand a sidebar with a picked item and make sure the item has scrolled into view.

Good call. I've added that in 1aae4fd.

**Comment by shaunandrews** [comment]
Somewhere along the way we got a horizontal scrollbar:

screen shot 2016-01-21 at 11 22 27 am

**Comment by shaunandrews** [comment]
When adding a list or tag, the iOS keyboard slides up. Perfect. But, if you then change your mind and tap the "X" icon to cancel, the keyboard doesn't hide:

screen shot 2016-01-21 at 11 29 21 am

**Comment by shaunandrews** [comment]
Outside of those above bugs, this looks/feels really good.

I wish the sidebar would remember it's states between sessions. Having them always collapse when I open a new tab is a little annoying.

File: client/reader/sidebar/expandable-heading.jsx [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * Internal dependencies
+ */
+import SidebarHeading from 'layout/sidebar/heading';
+import Gridicon from 'components/gridicon';
+import Count from 'components/count';
+
+const ExpandableSidebarHeading = ( { title, count, onClick } ) => (
+	<SidebarHeading onClick={ onClick }>
```
Comment (mtias) [inline client/reader/sidebar/expandable-heading.jsx]: Looking at this component I think we could make it a prop on `<SidebarHeading expandable>` and pass `Count` as a child only if you need to.

Comment (mtias) [inline client/reader/sidebar/expandable-heading.jsx]: And to clarify, I mean making this a general `layout/sidebar/heading` thing.

File: client/reader/sidebar/lists/index.jsx [hunk 13]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (mtias) [inline client/reader/sidebar/lists/index.jsx]: Just a reminder that by adding the `lists` folder you are creating a "component" namespace, so it should probably be `reader-lists` or `reader-sidebar-lists`.

File: client/reader/sidebar/tags/index.jsx [hunk 14]
```diff
@@ -0,0 +1,70 @@
+/**
```
Comment (mtias) [inline client/reader/sidebar/tags/index.jsx]: Same consideration as the one for `lists`.

**Comment by blowery** [comment]
This needs a rebase now too.

File: client/reader/sidebar/expandable-heading.jsx (hunk 12 above)
Comment (bluefuton) [inline client/reader/sidebar/expandable-heading.jsx]: Thanks @mtias. I'm definitely in favour of extracting this to `layout/sidebar`, but probably a separate PR would be best at this point.

File: client/reader/sidebar/lists/index.jsx (hunk 13 above)
Comment (bluefuton) [inline client/reader/sidebar/lists/index.jsx]: :+1:

File: client/reader/sidebar/tags/index.jsx (hunk 14 above)
Comment (bluefuton) [inline client/reader/sidebar/tags/index.jsx]: :+1:

**Comment by bluefuton** [comment]
Todo: 
- [ ] Find cause of horizontal scrollbar
- [ ] Fix iOS keyboard behaviour on cancel
- [ ] Prefix lists and tags folders with reader-sidebar-

**Comment by bluefuton** [comment]
Rebased.

**Comment by bluefuton** [comment]
Waiting on Redux state persistence (https://github.com/Automattic/wp-calypso/pull/2754) to retain open state between sessions.

**Comment by blowery** [comment]
I like it, works well!

Should we open the tag section if you click on a tag you're following in a post in the following stream? I don't _think_ so, but wanted do bring it up.

**Comment by blowery** [comment]
Should probably scroll the picked thing into view when it mounts or changes state?

**Comment by blowery** [comment]
I think this is good to go. Whatcha think @bluefuton ? Anything left to do?

**Comment by blowery** [comment]
Are there new stats we need to track?

**Comment by bluefuton** [comment]
> I think this is good to go. Whatcha think @bluefuton ? Anything left to do?

I think it'd good to get it out there, then address the "scroll the picked thing into view" and open/closed state persistence as enhancements later on.

> Are there new stats we need to track?

Yep, we could potentially add one for the open and close? Let me add that & then hopefully it's :ship: time :)

**Comment by bluefuton** [comment]
- [ ] Add stat for open and close of expandable menu

**Comment by blowery** [comment]
Rebased and squashed this down to author-driven commits. 

Looks like we lost the new display of tag names in the move?
screenshot

**Comment by bluefuton** [comment]
> Looks like we lost the new display of tag names in the move?

Sorted in 384eaf0 :+1:

**Comment by bluefuton** [comment]
Just found an IE11 CSS bug when expanding the add input:

https://cloudup.com/c_9Gix8SUYZ

**Pull request merged**